vm-device/src/device.rs

```diff
@@ -0,0 +1,110 @@
+// Copyright © 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+//! Handles routing to devices in an address space.
+use std::sync::{Arc, Mutex};
+use vm_memory::{GuestAddress, GuestUsize};
+
+/// Trait for devices with basic functions.
+#[allow(unused_variables)]
+pub trait Device: Send {
+    /// Get the device type.
+    fn device_type(&self) -> DeviceType;
+    /// Read from the guest physical address `addr` to `data`.
+    fn read(&mut self, addr: GuestAddress, data: &mut [u8], io_type: IoType);
+    /// Write `data` to the guest physical address `addr`.
+    fn write(&mut self, addr: GuestAddress, data: &[u8], io_type: IoType);
+    /// Set the allocated resource to device.
+    ///
+    /// This will be called by DeviceManager::register_device() to set
+    /// the allocated resource from the vm_allocator back to device.
+    fn set_resources(&mut self, res: &[IoResource], irq: Option<IrqResource>);
+}
+
+/// IO Resource type.
+#[derive(Debug, Copy, Clone)]
+pub enum IoType {
+    /// Port I/O resource.
+    Pio,
+    /// Memory I/O resource.
+    Mmio,
+    /// Non-exit physically backed mmap IO
+    PhysicalMmio,
+}
+
+/// Device resource information.
+#[derive(Debug, Copy, Clone)]
+pub struct IoResource {
+    /// Resource address.
+    pub addr: Option<GuestAddress>,
+    /// Resource size.
+    pub size: GuestUsize,
+    /// Resource type.
+    pub res_type: IoType,
+}
+
+impl IoResource {
+    /// Build a Resource struct.
+    pub fn new(addr: Option<GuestAddress>, size: GuestUsize, res_type: IoType) -> IoResource {
+        IoResource {
+            addr,
+            size,
+            res_type,
+        }
+    }
+}
+
+/// Legacy interrupt resource.
+pub struct IrqResource(pub Option<u32>);
```

Comment:
Maybe here we could do the same as we did with `IoResource`, by defining some `IrqType` types such as `PinBased` (or `Legacy`), `Msi` and `Msix`, wdyt?

Reply:
IMHO, msi / msix is not kind of resource that is allocated by vm-allocator.

Reply:
I understand, but even if the VMM is not the one allocating the MSI/MSI-X, it still need to choose which type of interrupt it wants to use for this device.
And you can also think about the fact that the device could be only PinBased capable even if the VMM ask for MSI, that's why the device should be able to advertise what type of IrqResource has been picked eventually, don't you think?

Reply:
@sboeuf Agree with you. The capability between VMM requesting and device supporting should been aligned in some way. Currently, I didn't see such things. :(
And I think we need another API  in vm-device to provide this information. 
For current cloud-hypervisor codes and latest PR #39, we can see the virtio registering logic is,
```
if interrupt_info.msi_capable {
    device_manager
                .register_device(
                 [...]
                 None, // no legacy pin should been requested
                )
    ...
} else {
    device_manager
                .register_device(
                 [...]
                 Some(IrqResource(None)),
                )
}
```